src/riak_kv_entropy_manager.erl

```diff
@@ -644,14 +664,52 @@ maybe_upgrade(State=#state{trees_version = VTrees}) ->
 check_exchanges_and_upgrade(State) ->
     case riak_kv_entropy_info:all_sibling_exchanges_complete() of
         true ->
-            lager:notice("Starting AAE hashtree upgrade"),
-            State#state{pending_version=0};
+            %% Now check nodes who havent reported success in the ETS table
+            {ok, Ring} = riak_core_ring_manager:get_my_ring(),
+            Nodes = riak_core_ring:all_members(Ring),
+            case check_remote_exchanges(Nodes,[]) of
+                complete ->
+                    lager:notice("Starting AAE hashtree upgrade"),
+                    State#state{pending_version=0};
+                _ ->
+                    State
+            end;
         _ ->
             State
     end.
 
+-spec check_remote_exchanges(list(), list()) -> complete | incomplete.
+check_remote_exchanges([], Acc) ->
```

Comment:
This looks like a lists:all call in disguise - if you take the check_remote_exchanges logic and split it into a few functions you end up with something like:

```erlang

check_all_remote_exchanges_complete(Nodes) ->
    lists:all(fun maybe_check_and_record_remote_exchange/1, Nodes).

maybe_check_and_record_remote_exchange(Node) ->
    case ets:lookup(?ETS, Node) of
         [{Node, true}] ->
             true;
         _ ->
             Result = check_remote_exchange(Node),
             ets:insert(?ETS, {Node, Result}),
             Result
     end.
```

This has a few advantages:
* Readability - it's clear what you're trying to accomplish.
* Bails on the first "false" answer
* Doesn't have to build/sort a list of results.
* We're trying to encourage the use of functions like lists:all, lists:map, etc. when they make sense, versus hand-rolling recursive functions on our own (or even using a foldl when you really mean map). This generally helps code maintainability.